### PR TITLE
Handling of multiple attribute values and LDAP entries for the 'attribute' query.

### PR DESCRIPTION
**Pull request by Ayanda-D**
Fixes: #16

### CODE REVIEW

**Comment by michaelklishin** [comment]
@Ayanda-D any chance we can add an integration test? [Our test setup](https://github.com/rabbitmq/rabbitmq-auth-backend-ldap/blob/master/README-tests.md) isn't great but we do have a test suite.

**Comment by Ayanda-D** [comment]
@michaelklishin sure, I'm having a look.

**Comment by michaelklishin** [comment]
@Dzol this primarily needs tests.

**Comment by Ayanda-D** [comment]
Tests for #16 are now covered.

**Comment by michaelklishin** [comment]
@Ayanda-D thank you. This no longer merges into `stable`, please resolve conflicts.

**Comment by Ayanda-D** [comment]
@michaelklishin sorted.

**Comment by michaelklishin** [comment]
Not assigning anyone as this will be reviewed by @dcorbacho, @hairyhum, and me.

**Comment by michaelklishin** [comment]
Seeding fails for me on this branch (but works on `stable`):

```
# ./example/setup.sh
#
# [snip]
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"

adding new entry "olcDatabase=bdb,cn=config"

adding new entry "cn=module,cn=config"

adding new entry "olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config"
ldap_add: No such object (32)
    matched DN: cn=config
```

**Comment by dcorbacho** [comment]
I get something different after run setup.sh and seed.sh, but the tests pass afterwards:

```
$ sh seed.sh 
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1

adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1

modifying entry "cn=module{1},cn=config"
ldap_modify: Type or value exists (20)
        additional info: modify/add: olcModuleLoad: value #0 already exists

adding new entry "olcOverlay={1}refint,olcDatabase={1}hdb,cn=config"

adding new entry "dc=example,dc=com"
ldap_add: Already exists (68)

adding new entry "ou=groups,dc=example,dc=com"
ldap_add: Already exists (68)

adding new entry "ou=vhosts,dc=example,dc=com"
ldap_add: Already exists (68)
```

File: src/rabbit_auth_backend_ldap.erl [hunk 1]
```diff
@@ -323,6 +323,36 @@ get_or_create_conn(IsAnon, Servers, Opts) ->
...
+        _                            -> get_attributes(AttrName, Rem)
+    end;
+get_attributes(AttrName, [_|Rem])    -> get_attributes(AttrName, Rem).
+
+%% Format multiple attribute values for logging
+format_multi_attr(Attrs) ->
+    format_multi_attr(io_lib:printable_list(Attrs), Attrs, []).
+
+format_multi_attr(true, Attrs, _Acc)       -> Attrs;
+format_multi_attr(_, [], Acc)              -> lists:flatten(Acc);
```
Comment (dcorbacho) [inline src/rabbit_auth_backend_ldap.erl]: Maybe all the `false` case could be replaced by `string:join(Attrs, "; ")` ?

File: src/rabbit_auth_backend_ldap.erl [hunk 2]
```diff
@@ -323,6 +323,36 @@ get_or_create_conn(IsAnon, Servers, Opts) ->
...
+format_multi_attr(F = false, [H|[]], Acc)  -> format_multi_attr(F, [], [H| Acc]);
+format_multi_attr(F = false, [H|Rem], Acc) ->
+  format_multi_attr(F, Rem, [Acc | ["; ", H]]).
+
+%% In case of multiple attributes, check for equality bi-directionally
+is_multi_attr_member(Str1, Str2) ->
```
Comment (dcorbacho) [inline src/rabbit_auth_backend_ldap.erl:366]: Equivalent to `lists:member(Str1, Str2) orelse lists:member(Str2, Str1)`

**Comment by hairyhum** [comment]
This PR is for `stable`, but contains `.schema` file. Looks like some leak from `master`.

**Comment by hairyhum** [comment]
@Ayanda-D according to commit history this PR is based on `master` and needs to be rebased or cherry-picked to be based `stable`.

File: src/rabbit_auth_backend_ldap.erl (hunk 1 above)
Comment (Ayanda-D) [inline src/rabbit_auth_backend_ldap.erl]: @dcorbacho yes they could. I'll update this. Thanks

File: src/rabbit_auth_backend_ldap.erl (hunk 2 above)
Comment (Ayanda-D) [inline src/rabbit_auth_backend_ldap.erl:366]: @dcorbacho true. I'll update. Thanks for this :+1:

**Comment by Ayanda-D** [comment]
@hairyhum hmm, okay cool, noted. I'll resolve this. thanks :+1:

**Comment by Ayanda-D** [comment]
Done with the suggested updates.

**Comment by dcorbacho** [comment]
@Ayanda-D The commit history still contains the .schema changes, could you delete them?

**Comment by hairyhum** [comment]
Looks fine to me.

**Comment by michaelklishin** [comment]
I cannot run the setup/seed script on this branch:

```
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"

adding new entry "olcDatabase=bdb,cn=config"

adding new entry "cn=module,cn=config"

adding new entry "olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config"
ldap_add: No such object (32)
    matched DN: cn=config
```

while it's OK on `stable`.

**Comment by Ayanda-D** [comment]
@michaelklishin this works just fine for me, & considering parent entries of `olcOverlay={0}memberof,olcDatabase={1}hdb,cn=config` are successfully added. I'm not reproducing this same error you're getting.

**Comment by dcorbacho** [comment]
I'm getting this:

```
ubuntu@ubuntu:~/rabbitmq-public-umbrella/deps/rabbitmq_auth_backend_ldap/example$ ./seed.sh 
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
adding new entry "cn=module,cn=config"
ldap_add: Other (e.g., implementation specific) error (80)
        additional info: <olcModuleLoad> handler exited with 1
```

**Comment by Ayanda-D** [comment]
@dcorbacho `./seed.sh` should already be executed as part of `./setup.sh`. no need to run it independently.

**Comment by dcorbacho** [comment]
@Ayanda-D right, then it works for me.

**Comment by hairyhum** [comment]
`./setup.sh` works fine for me on fresh installed `slapd-2.4.31-1` on Ubuntu 14.04.3 LTS

**Comment by michaelklishin** [comment]
OK, then the error I'm running into shouldn't prevent this from being merged.

**Pull request merged**